### PR TITLE
Assignment 3 @avanchu253

### PR DESCRIPTION
**Pull request by randypipkins**
Plus extra credit 2 completed

### CODE REVIEW

**Comment by jstanley6** [comment]
I will review this

**Comment by FenrisStargazer** [comment]
I will review this.

**Review by jstanley6 — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Review by FenrisStargazer — COMMENTED** [review]
**Assignment**
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ❌
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
_(I have no clue how to tell if it's <90% exactly, but it seems like you've tested everything.)_
I would say though that the test to make sure the joke doesn't contain Chuck Norris might be kinda iffy because what if the first joke it spits out just honestly doesn't contain it? Not like it's filtered out but like the API just returns a joke that doesn't contain it on the very first go. Then the filter isn't used and thus not tested... This probably doesn't really hurt anything but it's just something to think about.

**Fundamentals**
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
_( LangVersion is correct, but TargetFramework should be net 7.0)_
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌/✔??
_(I'm unsure if the way you did this is proper because "AssertEquals<T>" exactly like that didn't work for me)_
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Extra Credit**
Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)? ❌
The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string. ✔



When it runs it's cutting off the beginning several characters (about 9 or so?) of the joke, so it isn't printing properly. This also means some Chuck Norris jokes slip by the filter because some part of the name is cut off and thus not recognized.

**Comment by randypipkins** [comment]
@FenrisStargazer 
"For this assignment, always use Assert.AreEqual() (the generic version) ❌/✔??
(I'm unsure if the way you did this is proper because "AssertEquals< T >" exactly like it's shown didn't work for me)"

Not too sure as to why it is not working for you. Took the code and ran the test again a couple times and all of the tests passed fine, even after updating the target frame work to net 7.0.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+public class Jester
+{
+    public Jester(IJokeOutput jokeOutput, IJokeService jokeService)
+    {
+        JokeOutput = jokeOutput ?? throw new ArgumentNullException(nameof(jokeOutput));
+        JokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+    }
+
+    public IJokeOutput JokeOutput { get; }
+    public IJokeService JokeService { get; }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:14]:
```suggestion
    private IJokeOutput _JokeOutput;
    private IJokeService _JokeService;
```
If these don't need to be public, make them private. Right now, they are only being used within this class.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    public string? Joke { get; set; }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:16]: Keep this scoped to the TellJoke method, there's no reason you want to be able to access this outside of that method.

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 3]
```diff
@@ -2,14 +2,23 @@
...
-    public class JokeService
+    public class JokeService : IJokeService
     {
         private HttpClient HttpClient { get; } = new();
 
         public string GetJoke()
         {
-            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api").Result;
-            return joke;
+            Uri uri = new Uri("https://geek-jokes.sameerkumar.website/api?format=json%22");
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JokeService.cs:11]:
```suggestion
            Uri uri = new Uri("https://geek-jokes.sameerkumar.website/api?format=json");
```
Don't include this double quote in your URL, you aren't actually getting the json format back.

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 4]
```diff
@@ -2,14 +2,23 @@
...
+            string joke = HttpClient.GetStringAsync(uri).Result;
+            return JsonFormatStrip(joke);
+        }
+
+        private static string JsonFormatStrip(string jsonString)
+        {
+            string strippedString = jsonString.Remove(0, 0);
+            int index = strippedString.Length;
+            strippedString = strippedString.Remove(index, 0);
+            return strippedString;
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/JokeService.cs:21]: This is just stripping the joke string, not deserializing the JSON object, what you want to do is something like this:
`JokeInformation joke = JsonSerializer.Deserialize<JokeInformation>(json);`
```
public class JokeInformation
{
    [JsonPropertyName("joke")] 
    public string? Joke { get; set; }
}
```

**Pull request closed without merging**